### PR TITLE
LUCENE-10582: Fix merging of CollectionStatistics in CombinedFieldQuery

### PR DESCRIPTION
**Pull request by ywelsch**
CombinedFieldQuery does not properly combine distributed collection statistics, resulting in an IllegalArgumentException during searches.

Originally surfaced in this Elasticsearch issue: https://github.com/elastic/elasticsearch/issues/82817

### CODE REVIEW

File: lucene/sandbox/src/test/org/apache/lucene/sandbox/search/TestCombinedFieldQuery.java [hunk 1]
```diff
@@ -589,4 +589,52 @@ public SimScorer scorer(
       return new BM25Similarity().scorer(boost, collectionStats, termStats);
     }
   }
+
+  public void testDistributedCollectionStatistics() throws IOException {
```
Comment (jtibshirani) [inline lucene/sandbox/src/test/org/apache/lucene/sandbox/search/TestCombinedFieldQuery.java:593]: Small comment, maybe we could call this `testOverrideCollectionStatistics`? Lucene doesn't really have a native concept of "distributed collection statistics" (as far as I'm aware) and this test doesn't really use that concept anyway?

File: lucene/sandbox/src/test/org/apache/lucene/sandbox/search/TestCombinedFieldQuery.java [hunk 2]
```diff
@@ -589,4 +589,52 @@ public SimScorer scorer(
...
+        };
+    searcher.setSimilarity(new BM25Similarity());
+    CombinedFieldQuery query =
+        new CombinedFieldQuery.Builder()
+            .addField("f")
+            .addField("g")
+            .addTerm(new BytesRef(queryString))
+            .build();
+    // just check that search does not fail
+    searcher.search(query, 10);
```
Comment (jtibshirani) [inline lucene/sandbox/src/test/org/apache/lucene/sandbox/search/TestCombinedFieldQuery.java:634]: It'd be nice to assert something stronger here, to check that `CombinedFieldQuery` still works as expected when collection stats are overridden. Maybe we could compare the output of two query strategies like we do in `testCopyField`.

**Review by jtibshirani — COMMENTED** [review]
Thank you @ywelsch for debugging and fixing this!

File: lucene/sandbox/src/test/org/apache/lucene/sandbox/search/TestCombinedFieldQuery.java (hunk 1 above)
Comment (ywelsch) [reply lucene/sandbox/src/test/org/apache/lucene/sandbox/search/TestCombinedFieldQuery.java]: I used the term distributed as that's the use that is mentioned on the Javadocs of the collectionStatistics method. Fine to rename it here ([88b7f2c](https://github.com/apache/lucene/pull/910/commits/88b7f2ca8e44e554878a0c10f8ee6bfeb19e57d7)).

File: lucene/sandbox/src/test/org/apache/lucene/sandbox/search/TestCombinedFieldQuery.java (hunk 2 above)
Comment (ywelsch) [reply lucene/sandbox/src/test/org/apache/lucene/sandbox/search/TestCombinedFieldQuery.java]: I gave that a try in [88b7f2c](https://github.com/apache/lucene/pull/910/commits/88b7f2ca8e44e554878a0c10f8ee6bfeb19e57d7). Let me know what you think

File: lucene/sandbox/src/test/org/apache/lucene/sandbox/search/TestCombinedFieldQuery.java [hunk 3]
```diff
@@ -589,4 +589,97 @@ public SimScorer scorer(
...
+      }
+      w.addDocument(doc);
+    }
+
+    IndexReader reader = w.getReader();
+
+    int extraMaxDoc = randomIntBetween(0, 10);
+    int extraDocCount = randomIntBetween(0, extraMaxDoc);
+
+    int extraSumDocFreqA = extraDocCount + randomIntBetween(0, 10);
```
Comment (jtibshirani) [inline lucene/sandbox/src/test/org/apache/lucene/sandbox/search/TestCombinedFieldQuery.java:632]: I think it'd make more sense to have a single `sumDocFreq` here. This represents the number of unique term-document pairs, and we can't just add the values across different fields. In fact `CombinedFieldQuery` chooses to take a maximum of the `sumDocFreq`.

**Review by jtibshirani — COMMENTED** [review]
I left one comment, but other than that it looks good to me! Could you add an entry to CHANGES.txt too with your name?

**Review by jtibshirani — APPROVED** [review]

**Pull request merged**